### ISSUE TITLE
Scaffold components

### ISSUE THREAD

**Issue by yoshuawuyts** [choojs/choo-scaffold#2]
Once we settle on https://github.com/choojs/choo/issues/593, `choo-scaffold` should be able to create components too. This is probably the main reason why this module exists in the first place, as creating components requires a fair bit of boilerplate.

**Comment by YerkoPalma** [comment]
@choojs/trainspotters what would be a good boilerplate for components here?? I want to implement this.

Is something like this ok?

```js
var Component = require('choo/component')
var html = require('choo/html')

class component extends Component {
  constructor () {
    super()
  }

  createElement () {
    return html`
      <div>
      </div>
    `
  }

  update () {
    return false
  }
}

module.exports = component
```

**Comment by kareniel** [comment]
That's what I would go for personally.

**Comment by goto-bus-stop** [comment]
Same, I like that!

**Comment by nicknikolov** [comment]
I feel like update returning `false` by default might cause some mild headaches to people choo-ing for the first time? But might be fine.

**Comment by YerkoPalma** [comment]
There's no special reason to set it to false, so I could change it to true by default.

**Comment by ungoldman** [comment]
LGTM! I think defaulting to `createElement` accepting `state, emit` as parameters (and `update` too) is a small but important convention worth adopting.

**Comment by YerkoPalma** [comment]
Agree. So this might be a _better_ default?

```js
var Component = require('choo/component')
var html = require('choo/html')

class component extends Component {
  constructor () {
    super()
  }

  createElement (state, emit) {
    return html`
      <div>
      </div>
    `
  }

  update (state, emit) {
    return true
  }
}

module.exports = component
```

**Comment by tornqvist** [comment]
Assuming the component is to be rendered with the new component cache I think it'd make sense to reflect the signature imposed by the cache. It'd also help educate users on how to use the new component cache and encourage its use.

The constructor will have the component cache `id`, `state` and `emit` prefixed to the arguments. Also, with the introduction of choo/component the `components` property was added to the state. The intention of this was to have a shared space for all stateful things, including local component state, see reasoning for it [here](https://github.com/choojs/choo/pull/606#discussion_r156743091). @yoshuawuyts introduced the pattern of assigning a branch of the shared component state to `this.local` which I think is quite nice. 

```javascript
constructor (id, state, emit) {
  super(id)
  this.local = state.components[id] = {}
}
```

Idk if it'd make sense to assign `state` and `emit` to `this` in the constructor or if that's better left up to the user but either way there'd be no need to have them as default args to `createElement`.

**Comment by goto-bus-stop** [comment]
yeah using the API that `state.cache()` uses would be best :+1:

**Comment by YerkoPalma** [comment]
I'm ok with passing those arguments to the constructor, it will help a lot to folks like me :)

I think is up to the developer how those are used, that's why I wouldn't assign them to `this` in the constructor. Now, I would keep `state` in `createElement`, because if the state of the choo app change outside the component, we might want to keep it synced inside.

**Comment by YerkoPalma** [comment]
So, for now component boilerplate would be like this

```bash
$ choo-scaffold component custom-button
``` 

```js
var Component = require('choo/component')
var html = require('choo/html')

class CustomButton extends Component {
  constructor (id, state, emit) {
    super(id)
    this.local = state.components[id] = {}
  }

  createElement (state, emit) {
    return html`
      <div>
      </div>
    `
  }

  update (state, emit) {
    return true
  }
}

module.exports = CustomButton
```

if everyone is ok I will update the PR #6

**Comment by YerkoPalma** [comment]
PR updated 👍

**Comment by ungoldman** [comment]
This may be a little off-topic but I can't see why `state` and `emit` are getting passed to the Component constructor for the cache from looking at the cache source. AFAIK Nanocomponent isn't designed to do anything with state or emit until the first render, right? cc @bcomnes 

EDIT: sorry just saw the `this.local = state.components[id] = {}` line above.

Still confused on intended usage, how much component has to know about cache, etc., but this is off-topic.

**Comment by tornqvist** [comment]
@ungoldman The idea was to pass them to the constructor so that you *wouldn't* have to manually pass `state` and `emit` through the render method. I've been using it a lot as a way to share the component cache without having to pass it through the render call. I've also found that having `emit` attached to the component is very useful for wrapper components which only delegate callbacks to it's child components. This is what my component constructor boilerplate usually look like nowadays:

```javascript
constructor (id, state, emit) {
  super(id)
  this.emit = emit
  this.cache = state.cache
  this.local = state.components[id] = {}
}
```

**Comment by YerkoPalma** [comment]
@tornqvist you keep a cache reference in component to render other components? like [component container pattern](https://github.com/choojs/nanocomponent#nested-components-and-component-containers) described in nanocomponent docs?

**Comment by tornqvist** [comment]
That's right @YerkoPalma. Since the introduction of component cache in Choo core I only manually handle component instances when absolutely necessary and use `state.cache` otherwise. The fact that the constructor get the state (and hence, the cache) as an argument makes it very easy to have all components in the app at any depth share the same cache without much boilerplate at all.

**Comment by YerkoPalma** [comment]
Merged! but I can't publish :no_mouth:

**Comment by goto-bus-stop** [comment]
@YerkoPalma yosh added it to the org just now, you should be able to publish :)

**Comment by ungoldman** [comment]
@YerkoPalma you should be able to publish to npm, looks like you're on the access list: https://www.npmjs.com/package/choo-scaffold/access

**Comment by ungoldman** [comment]
haha woops looks like @goto-bus-stop and yosh already took care of it at the exact time I looked 👀

**Comment by YerkoPalma** [comment]
haha nice! just got published as `1.2.0` :tada:

thank you all :)
